### PR TITLE
Add MLX backend support for Gemma 4 31B

### PR DESCRIPTION
**Pull request by mergennachin**
Adds Apple Silicon (MLX) backend for the Gemma 4 31B-IT model. The same quantized checkpoint works for both CUDA and MLX — backend-specific packing happens at load time.

Key changes:

- MLX packer converts Int4Tensor → IntxUnpackedToInt8Tensor for MLX's quantized linear fusion
- Source transforms replace PyTorch ops with mlx.rope, mlx.kv_cache_update, mlx.custom_sdpa for optimized Metal kernels
- Proportional partial RoPE (full-attention layers) passes 1D frequencies to mlx.rope with dims=rotary_dim, fixing the C++ runtime to pass base=nullopt when freqs is provided
- Single-method export with dynamic seq_len and host-side sampling
- C++ runner supports both backends via #ifdef, using shared logits_to_token for MLX sampling
- Last-logits-only optimization: lm_head always runs on last position only, removing the full-logits codepath entirely

Nothing in the CUDA backend code itself. The CUDA-side changes are in the shared model/runner code:

- model.py: forward() now always does last-logits-only and temperature is required (no None path). Affects both CUDA and MLX.
- sampler.py: Removed temperature=None passthrough.
- main.cpp: Unified temp_val clamping before the #ifdef. CUDA path behavior unchanged.
- inference.py: Default temperature changed from 0.0 to 0.8 to match C++ runner default.

On my 32GB RAM M1 macbook pro 
```
(executorch_dev) mnachin@mnachin-mbp executorch % cmake-out/examples/models/gemma4_31b/gemma4_31b_runner --model_path ~/repos/models/gemma-4-31B-it-HQQ-INT4/model.pte --tokenizer_path ~/repos/models/gemma-4-31B-it-HQQ-INT4/tokenizer.json --prompt "Write a short joke about saving RAM." --max_new_tokens 128
I tokenizers:regex.cpp:27] Registering override fallback regex
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1779218557.174278 43844526 re2.cc:237] Error parsing '((\<pad\>|ool\|\>1\x00\x00\
                                                                                             �\<t|respo|\<tool_call\|\>|\<bos\>|\<\|tool_response\>|\<\|think\|\>|\x0...': invalid UTF-8
I tokenizers:re2_regex.cpp:27] Re2 failed to compile regex: ((\<pad\>|ool\|\>1\x00\x00\
                                                                                       �\<t|respo|\<tool_call\|\>|\<bos\>|\<\|tool_response\>|\<\|think\|\>|\x00\x00\\\<|\<tool_response\|\>|\<mask\>|\<\|\"\|\>|all\|\>j\x00\x00\\|\<channel\|\>|\<\|turn\>|\<turn\|\>|\<\|image\>|\<\|$
I tokenizers:regex_lookahead.cpp:27] Creating PCRE2 regex
I tokenizers:pcre2_regex.cpp:48] PCRE2 UTF-8 validation failed at offset 27: UTF-8 error: byte 2 top bits not 0x80. Retrying without UTF flags.
Loading model...
Prompt tokens: 24
Why did the programmer get kicked out of the library?

He kept trying to free the memory.<turn|>
PyTorchObserver {"prefill_token_per_sec":7.56859,"decode_token_per_sec":2.09161,"prompt_tokens":24,"generated_tokens":20,"model_load_start_ms":1779218556804,"model_load_end_ms":1779218560048,"inference_start_ms":1779218560052,"inference_end_ms":1779218572785,"prompt_eval_end_ms":1779218563223,"first_token_ms":1779218563223,"aggregate_sampling_time_ms":0,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
```

### CODE REVIEW

File: examples/models/gemma4_31b/main.cpp [hunk 1]
```diff
@@ -106,6 +107,25 @@ static uint64_t read_token(const executorch::aten::Tensor& output) {
   return static_cast<uint64_t>(llrintf(val));
 }
 
+// Greedy argmax over the last token's logits (MLX path).
+static uint64_t argmax_last_token(const executorch::aten::Tensor& logits) {
```
Comment (metascroy) [inline examples/models/gemma4_31b/main.cpp]: Why no temperature?

File: examples/models/gemma4_31b/quant/pack_mlx.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    only accepts group_size in {32, 64, 128}.
+    """
+    from torchao.quantization import IntxUnpackedToInt8Tensor
+    from torchao.quantization.quantize_.workflows.int4.int4_tensor import Int4Tensor
+
+    w = weights["weight"]
+    if isinstance(w, Int4Tensor):
+        raise ValueError(
+            "Only 8-bit embedding quantization is supported on MLX. "
+            "INT4 does not implement the embedding op."
```
Comment (metascroy) [inline examples/models/gemma4_31b/quant/pack_mlx.py]: int4 embedding should work on mlx?

File: examples/models/gemma4_31b/quant/pack_mlx.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        f"(or a multiple thereof), got {gs}"
+    )
+
+
+def _regroup_intx(w: torch.Tensor, new_gs: int) -> torch.Tensor:
+    """Regroup an ``IntxUnpackedToInt8Tensor`` to a finer group_size."""
+    from torchao.quantization import IntxUnpackedToInt8Tensor
+
+    old_gs = w.block_size[-1]
+    repeat_factor = old_gs // new_gs
```
Comment (metascroy) [inline examples/models/gemma4_31b/quant/pack_mlx.py:98]: Shouldn't we error out if old_gs doesn't divide new_gs?

**Comment by mergennachin** [comment]
@claude Review this PR

File: examples/models/gemma4_31b/quant/tests/test_pack_mlx.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                config = QuantConfig(
+                    bits=4, group_size=gs, symmetric=True, method="min_max"
+                )
+                int4_w = quantize_weight(
+                    torch.randn(64, 256, dtype=torch.bfloat16), config
+                )
+                intx_w = _int4_to_intx_unpacked(int4_w)
+                self.assertEqual(intx_w.shape, torch.Size([64, 256]))
+
+    def test_matmul_approximates_original(self):
```
Comment (metascroy) [inline examples/models/gemma4_31b/quant/tests/test_pack_mlx.py:86]: Do we have a test case that covers comparing regrouped linear with original?

File: examples/models/gemma4_31b/quant/pack_mlx.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    w = weights["weight"]
+    if isinstance(w, Int4Tensor):
+        w = _int4_to_intx_unpacked(w)
+    elif not isinstance(w, IntxUnpackedToInt8Tensor):
+        raise ValueError(f"Unsupported weight type: {type(w).__name__}")
+    module.weight = nn.Parameter(w, requires_grad=False)
+
+
+def pack_embedding_for_mlx(module: nn.Module, weights: dict[str, torch.Tensor]) -> None:
```
Comment (metascroy) [inline examples/models/gemma4_31b/quant/pack_mlx.py]: Is there a reason why we need separate functions for linear and embedding?

It looks like both functions do the same conversion, except embedding also has the regrouping logic (which linear could also use)?

**Review by metascroy — APPROVED** [review]

File: examples/models/gemma4_31b/mlx_source_transformations.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    import types
+
+    def _mlx_model_forward(
+        self, tokens: torch.Tensor, input_pos: torch.Tensor
+    ) -> torch.Tensor:
+        x = self.embed_tokens(tokens) * self.embed_normalizer
+        for layer in self.layers:
+            x = layer(x, input_pos)
+        x = self.norm(x)
+        last = self.lm_head(x[:, -1, :]).float()
```
Comment (metascroy) [inline examples/models/gemma4_31b/mlx_source_transformations.py:143]: Do we need to cast here? Does torch.tanh not use higher precision internally?

Comment (mergennachin) [reply examples/models/gemma4_31b/mlx_source_transformations.py:143]: The .float() cast is needed. Without it, the tanh would run in bf16 (input dtype)
